File: sea-shell-service-spring-web-flux/sea-shell-rest-service/src/main/java/org/jesperancinha/shell/webflux/service/fork/SeaShellTopRecursiveTask.java
package org.jesperancinha.shell.webflux.service.fork;

import lombok.AllArgsConstructor;
import lombok.Builder;
import org.jesperancinha.shell.webflux.data.SeaShellCostumeDto;
import org.jesperancinha.shell.webflux.repo.ShellTopRepository;
import org.jesperancinha.shell.webflux.service.SeaShellConverter;

import java.util.concurrent.RecursiveTask;

@Builder
@AllArgsConstructor
public class SeaShellTopRecursiveTask extends RecursiveTask<SeaShellCostumeDto> {

    private final ShellTopRepository topRepository;
    private final SeaShellCostumeDto costumeDto;

    @Override
    protected SeaShellCostumeDto compute() {
        costumeDto.setTopDto(SeaShellConverter.toTopDto(topRepository.findTopByIdBlock(costumeDto.getTopId())));
        return costumeDto;
    }
}

File: sea-shell-service-spring-web-flux/src/main/kotlin/org/jesperancinha/shell/webflux/service/fork/SeaShellTopRecursiveTask.kt
package org.jesperancinha.shell.webflux.service.fork

import lombok.AllArgsConstructor
import lombok.Builder
import org.jesperancinha.shell.webflux.data.SeaShellCostumeDto
import org.jesperancinha.shell.webflux.repo.ShellTopRepositoryImpl
import org.jesperancinha.shell.webflux.service.SeaShellConverter
import java.util.concurrent.RecursiveTask

@Builder
@AllArgsConstructor
class SeaShellTopRecursiveTask(
    val topRepository: ShellTopRepositoryImpl,
    val costumeDto: SeaShellCostumeDto
) : RecursiveTask<SeaShellCostumeDto>() {

    override fun compute(): SeaShellCostumeDto =
        costumeDto.copy(topDto = costumeDto.topId?.let {
            topRepository.findTopByIdBlock(it)
        }?.let { SeaShellConverter.toTopDto(it) })
}